src/protocol/sort/bit_permutation.rs

```diff
@@ -112,4 +117,40 @@ mod tests {
 
         validate_list_of_shares(EXPECTED, &result);
     }
+
+    #[tokio::test]
+    pub async fn test_bit_permutation_malicious() {
+        // With this input, for stable sort we expect all 0's to line up before 1's.
+        // The expected sort order is same as expected_sort_output.
+        const INPUT: &[u128] = &[1, 0, 1, 0, 0, 1, 0];
+        const EXPECTED: &[u128] = &[4, 0, 5, 1, 2, 6, 3];
+
+        let world = make_world(QueryId);
+        let [ctx0, ctx1, ctx2] = make_malicious_contexts::<Fp31>(&world);
+        let mut rand = StepRng::new(100, 1);
+
+        let mut shares = [
+            Vec::with_capacity(INPUT.len()),
+            Vec::with_capacity(INPUT.len()),
+            Vec::with_capacity(INPUT.len()),
+        ];
+        for i in INPUT {
+            let share = share_malicious(Fp31::from(*i), &mut rand);
```

Comment:
How can you share malicious without knowing the value of `r`?

Reply:
Ugh.. I missed this when `share_malicious()` was added to the codebase... I'll fix it in another diff.